### PR TITLE
Link the kernel at 0xFFFF_8000_0000_0000.

### PR DESCRIPTION
**Pull request by dancrossnyc**
Link the kernel at the beginning of the upper
part of the canonical address space.  Lots of
other miscellaneous fixes.

Signed-off-by: Dan Cross <cross@gajendra.net>

### CODE REVIEW

File: sys/src/9/amd64/kernel.ld [hunk 1]
```diff
@@ -1,78 +1,59 @@
...
+	 * start of the kernel for critical structures.
+	 */
+	. = 0xffff800000100000;
 
-	.bootstrap : {
+	PROVIDE(boottext = .);
+	.text.boot : ALIGN(4096) {
 		*(.boottext .bootdata)
+		. = ALIGN(4096);
+		PROVIDE(eboottext = .);
```
Comment (gmacd) [inline sys/src/9/amd64/kernel.ld:19]: are eboottext and esys used anywhere?  I see no reference to them

File: sys/src/9/amd64/kernel.ld [hunk 2]
```diff
@@ -1,78 +1,59 @@
...
+		PROVIDE(esys = .);
 	}
 
-	/* Link the main kernel for the space after entry + KERN_LOAD_ADDR.  We'll
-	 * still load it adjacent in physical memory */
-	. += KERN_LOAD_ADDR;
-
-	.text : AT(ADDR(.text) - KERN_LOAD_ADDR) {
-		*(.text .stub .text.* .gnu.linkonce.t.*)
+	PROVIDE(text = .);
```
Comment (gmacd) [inline sys/src/9/amd64/kernel.ld:24]: don't think text is actually used anywhere?  hard to tell

**Review by gmacd — APPROVED** [review]

File: sys/src/9/amd64/kernel.ld (hunk 2 above)
Comment (dancrossnyc) [reply sys/src/9/amd64/kernel.ld:24]: No, but it's useful for debugging: it shows up in a symbol dump, and we can use it to validate that the text segment starts at e.g. KTZERO.

File: sys/src/9/amd64/kernel.ld (hunk 1 above)
Comment (dancrossnyc) [reply sys/src/9/amd64/kernel.ld:19]: No, same thing as `text`: they appear in a symbol dump for debugging.

File: bootstrap.sh [hunk 3]
```diff
@@ -2,44 +2,47 @@
 
 set -e
 
-HOSTBIN=$(go env GOHOSTOS)_$(go env GOHOSTARCH)/bin
-export GOBIN=$(pwd)/$HOSTBIN
+export GOBIN=$(pwd)/$(go env GOHOSTOS)_$(go env GOHOSTARCH)/bin
```
Comment (rminnich) [inline bootstrap.sh:5]: is this a regression? You might need to do a git pull --rebase, not sure.

File: sys/src/9/amd64/archamd64.c [hunk 4]
```diff
@@ -192,13 +192,16 @@ cpuidhz(uint32_t *info0, uint32_t *info1, CpuHypervisor hypervisor)
 //print("msr 2a is 0x%x >> 22 0x%x\n", rdmsr(0x2a), rdmsr(0x2a)>>22);
 			break;
 		case 0x000306a0:		/* i7,5,3 3xxx */
+		case 0x000306f0:		/* i7,5,3 5xxx and Xeon dev machine */
```
Comment (rminnich) [inline sys/src/9/amd64/archamd64.c:195]: Looking at this, I think a rebase might be a good idea. This change was done earlier?

File: sys/src/9/amd64/l64vsyscall.S [hunk 5]
```diff
@@ -73,15 +73,16 @@ syscallentry:
 	MOVQ	%rsp, %rsi 				/* Ureg* */
 	// system call number is in %rax, as per linux.
 	movq	%rax, %rdi
-	xorq	%rax, %rax
-	pushq	%rax
+	xorl	%eax, %eax
```
Comment (rminnich) [inline sys/src/9/amd64/l64vsyscall.S:76]: why not xorq?

File: sys/src/9/amd64/l64vsyscall.S [hunk 6]
```diff
@@ -73,15 +73,16 @@ syscallentry:
...
+	pushq	$2
 	popfq			/* clear all flags. is there something else we should clear too? */
-	movq	$0, %rbp	/* stack traces end here */
+	pushq	$0
```
Comment (rminnich) [inline sys/src/9/amd64/l64vsyscall.S:79]: it might be best if this were a separate commit? What is this fixing?

File: sys/src/9/amd64/memory.c [hunk 7]
```diff
@@ -13,18 +13,101 @@
...
+		conf.npage += conf.mem[cx].npage;
+		DBG("cm %d: addr %#llx npage %lu\n",
+			cx, conf.mem[cx].base, conf.mem[cx].npage);
+		cx++;
+	}
+	mmukflushtlb();
+
+	/*
+	 * Fill in more legacy conf data.
+	 * This is why I hate Plan 9.
```
Comment (rminnich) [inline sys/src/9/amd64/memory.c:68]: This is a jmk comment and someday I'd like to figure out how to remove it.

File: sys/src/9/amd64/syscall.c [hunk 8]
```diff
@@ -532,21 +532,14 @@ sysprocsetup(Proc* p)
...
-// But it is the right value ...
-#define STACKPAD 1 /* for return PC? */
-	/*
-	 * Add STACKPAD*BY2SE to the stack to account for
-	 *  - the return PC
-	 *  (NOT NOW) - trap's arguments (syscallnr, ureg)
-	 */
-	child->sched.sp = PTR2UINT(child->kstack+KSTACK-((sizeof(Ureg)+STACKPAD*BY2SE)));
-	child->sched.pc = PTR2UINT(sysrforkret);
+	char *cureg;
```
Comment (rminnich) [inline sys/src/9/amd64/syscall.c:535]: should this be a separate PR; I'd like to know why you were able to remove STACKPAD.

**Review by rminnich — DISMISSED** [review]
There's a ton of good tidying up in here in addition to the basic improvement. Consider a rebase and also if you want to break out some of these into multiple PRs.

File: sys/src/9/amd64/archamd64.c (hunk 4 above)
Comment (dancrossnyc) [reply sys/src/9/amd64/archamd64.c:195]: The case statement moved from a place where it would try MSR 0xCD to here, so that it won't try that (which cased a GPF on my dev machine).

File: bootstrap.sh (hunk 3 above)
Comment (dancrossnyc) [reply bootstrap.sh:5]: Oh that's weird: I had rebased and then `git checkout --theirs -- bootstrap.sh` but it didn't take the most recent changes.

File: sys/src/9/amd64/l64vsyscall.S (hunk 5 above)
Comment (dancrossnyc) [reply sys/src/9/amd64/l64vsyscall.S:76]: It's unnecessary and a longer instruction (e.g., more instruction bytes). xorl will clear the upper 32 bits of the destination register anyway; this is the form recommended by the Intel optimization manual, which I've standardized on everywhere else as a matter of course (e.g., not just for optimized sections of code).

File: sys/src/9/amd64/l64vsyscall.S (hunk 6 above)
Comment (dancrossnyc) [reply sys/src/9/amd64/l64vsyscall.S:79]: This was just wrong; removed.

File: sys/src/9/amd64/memory.c (hunk 7 above)
Comment (dancrossnyc) [reply sys/src/9/amd64/memory.c:68]: Removing this will require a bit more further study, because plan9 made naive assumptions about the amount of virtual address space available as a proxy for allocatable RAM.

File: sys/src/9/amd64/syscall.c (hunk 8 above)
Comment (dancrossnyc) [reply sys/src/9/amd64/syscall.c:535]: Sure; reverted for now.

File: sys/src/9/amd64/apic.c [hunk 9]
```diff
@@ -419,10 +418,9 @@ apicsipi(int apicno, uintmem pa)
 	apicrput(Iclo, DSnone|TMlevel|MTir);
 	millidelay(10);
 
-	crlo = DSnone|TMedge|MTsipi|((uint32_t)pa/(4*KiB));
-	for(i = 0; i < 2; i++){
+	for(int i = 0; i < 2; i++){
 		apicrput(Ichi, crhi);
-		apicrput(Iclo, crlo);
+		apicrput(Iclo, DSnone|TMedge|MTsipi|((uint32_t)pa/(4*KiB)));
```
Comment (rminnich) [inline sys/src/9/amd64/apic.c:423]: 4 * KiB -> PGSZ?

File: sys/src/9/amd64/archamd64.c [hunk 10]
```diff
@@ -192,13 +192,16 @@ cpuidhz(uint32_t *info0, uint32_t *info1, CpuHypervisor hypervisor)
...
 		case 0x000206c0:		/* i7,5,3 4xxx */
 		case 0x00050650:		/* i9 7900X */ // Moved here by RGM 10/27/2020
 		case 0x000506e0:		/* i7,5,3 6xxx */
 			// reading msr 0xcd gets a GPF on this CPU.
-			// per the coreboot irc:
-			// <icon[m]> rminnich: if you need the base for the core's clock multiplier, it's 100MHz since sandybridge
-			// Which, going by the Good Book (35-46 volume 3C) is index 5.
+			// From Ron on coreboot irc:
```
Comment (rminnich) [inline sys/src/9/amd64/archamd64.c:200]: I was just confused about the comment change: I think that's in upstream. But now I understand.

File: sys/src/9/amd64/dat.h [hunk 11]
```diff
@@ -225,10 +224,7 @@ enum
 struct ICC
 {
 	/* fn is kept in its own cache line */
-	union{
-		void	(*fn)(void);
-		unsigned char	_ln1_[ICCLNSZ];
-	};
+	alignas(ICCLNSZ) void	(*fn)(void);
```
Comment (rminnich) [inline sys/src/9/amd64/dat.h:227]: This works on clang too right? This is kind of interesting trend as we move further away from ever running under Ken C again (I'm fine with that :-)) unless Ken C gets some upgrades. This is way clearer than the original as to intent ...

File: sys/src/9/amd64/dat.h [hunk 12]
```diff
@@ -354,69 +351,88 @@ struct Stackframe
...
  */
 struct Sys {
-	unsigned char	machstk[MACHSTKSZ];
+	alignas(4096) unsigned char machstk[MACHSTKSZ];
 
-	PTE	pml4[PTSZ/sizeof(PTE)];	/*  */
-	PTE	pdp[PTSZ/sizeof(PTE)];
-	PTE	pd[PTSZ/sizeof(PTE)];
-	PTE	pt[PTSZ/sizeof(PTE)];
+	PTE	ipml4[PTSZ/sizeof(PTE)];	// Only used very early in boot
```
Comment (rminnich) [inline sys/src/9/amd64/dat.h:370]: is PTSZ / sizeof(PTE) worth a define at some point?

File: sys/src/9/amd64/devacpi.c [hunk 13]
```diff
@@ -517,45 +517,31 @@ static void *sdtmap(uintptr_t pa, size_t want, size_t *n, int cksum)
 {
 	Sdthdr *sdt;
 	Acpilist *p;
+
 	if(v)print("sdtmap %p\n", (void *)pa);
 	if (!pa) {
 		print("sdtmap: nil pa\n");
 		return nil;
 	}
+	sdt = KADDR(pa);
```
Comment (rminnich) [inline sys/src/9/amd64/devacpi.c:526]: here we really start to see the big value add of this change.

File: sys/src/9/amd64/fns.h [hunk 14]
```diff
@@ -29,12 +29,6 @@ void	acmodeset(int);
 void	archfmtinstall(void);
 void	archidle(void);
 int	archmmu(void);
-int	asmfree(uintmem, uintmem, int);
```
Comment (rminnich) [inline sys/src/9/amd64/fns.h:32]: I will not miss asm.

File: sys/src/9/amd64/l64vsyscall.S (hunk 5 above)
Comment (rminnich) [reply sys/src/9/amd64/l64vsyscall.S:76]: fantastic.

**Review by rminnich — APPROVED** [review]
I'm fine as is, my comments are nits.

File: sys/src/9/amd64/apic.c (hunk 9 above)
Comment (rminnich) [reply sys/src/9/amd64/apic.c:423]: I mean PTSZ I guess?

File: sys/src/9/amd64/dat.h (hunk 12 above)
Comment (dancrossnyc) [reply sys/src/9/amd64/dat.h:370]: Probably, but I think best deferred to a followup change.

File: sys/src/9/amd64/dat.h (hunk 11 above)
Comment (dancrossnyc) [reply sys/src/9/amd64/dat.h:227]: It does indeed work with clang. It's actually a C11 thing.

File: sys/src/9/amd64/archamd64.c (hunk 10 above)
Comment (dancrossnyc) [reply sys/src/9/amd64/archamd64.c:200]: Ah yes; I just reformatted it.

**Comment by dancrossnyc** [comment]
> There's a ton of good tidying up in here in addition to the basic improvement. Consider a rebase and also if you want to break out some of these into multiple PRs.

Definitely did a rebase, but it's too entwined to reasonably separate out.

**Review by rminnich — APPROVED** [review]
